File: Kaiser/test/cc/game/kaiser/core/KaiserTest.java
package cc.game.kaiser.core;

import java.io.File;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.Arrays;

import cc.game.kaiser.ai.PlayerBot;
import cc.lib.game.Utils;
import junit.framework.TestCase;

public class KaiserTest extends TestCase {

    @Override
    public void setUp() {
        Kaiser.DEBUG_ENABLED = true;
        System.out.println("Start Test: " + getName());
    }
    
    public void testSerializeTeam() throws Exception {
        Team team = new Team();
        team.bid = Bid.NO_BID;
        team.name = "scrubs";
        team.roundPoints = 10000;
        team.totalPoints = 98320982;
        
        //PrintWriter out = new PrintWriter(System.out);
        StringWriter out = new StringWriter();
        team.serialize(new PrintWriter(out));
        System.out.println(out.toString());
        
        Team t2 = new Team();
        t2.deserialize(out.toString());
        
        assertEquals(t2.totalPoints, team.totalPoints);
        assertEquals(t2.roundPoints, team.roundPoints);
        assertEquals(t2.name, team.name);
        assertEquals(t2, team);
    }
    
    public void testKaiser() throws Exception {
        boolean saved = false;
        Kaiser k = new Kaiser();
        k.setPlayer(0, new PlayerBot("Joe Blow"));
        k.setPlayer(1, new PlayerBot("Pat Robertson"));
        k.setPlayer(2, new PlayerBot("Bobbing for Apples"));
        k.setPlayer(3, new PlayerBot("Sam crow"));
        while (k.getState() != State.GAME_OVER) {
            if (k.getState() == State.NEW_ROUND && !saved && Math.random() * 10 < 2) {
                k.saveToFile(new File("resources/kaisertest.txt"));
                saved = true;
            }
            k.runGame();
        }
    }

    public void testRestoreGame() throws Exception {
        Kaiser.DEBUG_ENABLED = true;
        Utils.DEBUG_ENABLED = false;
        Kaiser k = new Kaiser();
        k.setPlayer(0, new PlayerBot("Joe Blow"));
        k.setPlayer(1, new PlayerBot("Pat Robertson"));
        k.setPlayer(2, new PlayerBot("Bobbing for Apples"));
        k.setPlayer(3, new PlayerBot("Sam crow"));
        while (k.getState() != State.GAME_OVER) {
            k.saveToFile(new File("resources/test.txt"));
            k = new Kaiser();
            k.loadFromFile(new File("resources/test.txt"));
            k.runGame();
            System.out.println("State after run: " + k.getState());
        }
    }

    public void testRestoreGame2() throws Exception {
        Kaiser k = new Kaiser();
        k.loadFromFile(new File("resources/savegame.txt"));
        Player p = k.getPlayer(0);
        assertEquals("Chris", p.getName());
        assertEquals(7, p.getNumCards());
        assertEquals(1, k.getTeam(1).getRoundPoints());
        assertEquals(10, k.getTeam(1).getTotalPoints());

        int cnt = 1000;
        while (cnt-- > 0 && k.getState() != State.GAME_OVER) {
            k.runGame();
        }
        assertTrue(k.getState() == State.GAME_OVER);
    }
    
    public void testSerialize() throws Exception {
        
        Kaiser k = new Kaiser();
        k.loadFromFile(new File("resources/kaisertest.txt"));
        while (k.getState() != State.GAME_OVER) {
            k.runGame();
        }
    }
    
    public void testGetTrickWinnerIndex() {

        Card aceClubs = new Card(Rank.ACE, Suit.CLUBS);
        Card kngDmnds = new Card(Rank.KING, Suit.DIAMONDS);
        Card qunHrts = new Card(Rank.QUEEN, Suit.HEARTS);
        Card jckSpds = new Card(Rank.JACK, Suit.SPADES);
        Card jckHrts = new Card(Rank.JACK, Suit.HEARTS);
        Card jckClubs = new Card(Rank.JACK, Suit.CLUBS);
        Card jckDmnds = new Card(Rank.JACK, Suit.DIAMONDS);
        Card qunDmnds = new Card(Rank.QUEEN, Suit.DIAMONDS);
        Card tenDmnds = new Card(Rank.TEN, Suit.DIAMONDS);
        
        // test when all the same, no trump
        doTestTrick(new Card[] { jckSpds, jckHrts, jckClubs, jckDmnds}, Suit.NOTRUMP, 0, jckSpds);
        doTestTrick(new Card[] { jckSpds, jckHrts, jckClubs, jckDmnds}, Suit.NOTRUMP, 1, jckHrts);
        doTestTrick(new Card[] { jckSpds, jckHrts, jckClubs, jckDmnds}, Suit.NOTRUMP, 2, jckClubs);
        doTestTrick(new Card[] { jckSpds, jckHrts, jckClubs, jckDmnds}, Suit.NOTRUMP, 3, jckDmnds);
        
        // test when all the same, has trump
        for (int i=0 ;i<4; i++) {
            doTestTrick(new Card[] { jckSpds, jckHrts, jckClubs, jckDmnds}, Suit.SPADES, i, jckSpds);
            doTestTrick(new Card[] { jckSpds, jckHrts, jckClubs, jckDmnds}, Suit.HEARTS, i, jckHrts);
            doTestTrick(new Card[] { jckSpds, jckHrts, jckClubs, jckDmnds}, Suit.CLUBS, i, jckClubs);
            doTestTrick(new Card[] { jckSpds, jckHrts, jckClubs, jckDmnds}, Suit.DIAMONDS, i, jckDmnds);
        }
        
        // test when all different, no trump
        doTestTrick(new Card[] { aceClubs, kngDmnds, qunHrts, jckSpds }, Suit.NOTRUMP, 0, aceClubs);
        doTestTrick(new Card[] { aceClubs, kngDmnds, qunHrts, jckSpds }, Suit.NOTRUMP, 1, kngDmnds);
        doTestTrick(new Card[] { aceClubs, kngDmnds, qunHrts, jckSpds }, Suit.NOTRUMP, 2, qunHrts);
        doTestTrick(new Card[] { aceClubs, kngDmnds, qunHrts, jckSpds }, Suit.NOTRUMP, 3, jckSpds);
        
        doTestTrick(new Card[] { aceClubs, qunHrts, jckSpds, jckHrts }, Suit.NOTRUMP, 0, aceClubs);
        doTestTrick(new Card[] { aceClubs, qunHrts, jckSpds, jckHrts }, Suit.NOTRUMP, 1, qunHrts);
        doTestTrick(new Card[] { aceClubs, qunHrts, jckSpds, jckHrts }, Suit.NOTRUMP, 2, jckSpds);
        doTestTrick(new Card[] { aceClubs, qunHrts, jckSpds, jckHrts }, Suit.NOTRUMP, 3, qunHrts);
        
        // test when all different, has trump
        Suit [] suits = new Suit []{ Suit.CLUBS, Suit.DIAMONDS, Suit.HEARTS, Suit.SPADES };
        Card [] cards = new Card[] { aceClubs, kngDmnds, qunHrts, jckSpds };
        int index = 0;
        for (Suit trump: suits) {
        
            doTestTrick(cards, trump, 0, cards[index]);
            doTestTrick(cards, trump, 1, cards[index]);
            doTestTrick(cards, trump, 2, cards[index]);
            doTestTrick(cards, trump, 3, cards[index]);
            index++;
        }        
        
        // test all the same suit, no trump
        for (int i=0; i<4; i++)
            doTestTrick(new Card[] { kngDmnds, qunDmnds, jckDmnds, tenDmnds}, Suit.NOTRUMP, i, kngDmnds);

        for (int i=0; i<4; i++)
            doTestTrick(new Card[] { kngDmnds, qunDmnds, jckDmnds, tenDmnds}, Suit.DIAMONDS, i, kngDmnds);

        for (int i=0; i<4; i++)
            doTestTrick(new Card[] { kngDmnds, qunDmnds, jckDmnds, tenDmnds}, Suit.SPADES, i, kngDmnds);

    }
    
    void doTestTrick(Card [] trick, Suit trump, int startIndex, Card expected) {
        int index = Kaiser.getTrickWinnerIndex(trump, trick, startIndex);
        assertEquals(expected, trick[index]);
    }
 /*   
    public void testParseNextLineElement() {
        String line = "PLAYER:0 cc.game.kaiser.swing.SwingPlayerUser \"Chris\"";
        
        assertEquals("PLAYER", Kaiser.parseNextLineElement(line, 0));
        assertEquals("0", Kaiser.parseNextLineElement(line, 1));
        assertEquals("cc.game.kaiser.swing.SwingPlayerUser", Kaiser.parseNextLineElement(line, 2));
        assertEquals("Chris", Kaiser.parseNextLineElement(line, 3));
      
    }
   */ 
    public void test_getBestTrumpOptions() {
        
        Kaiser k = new Kaiser();
        for (int i=0; i<4; i++) {
            Card [] hand = new Card[] { k.getCard(Rank.ACE, Suit.values()[i]) };
        
            Suit [] suits = Kaiser.getBestTrumpOptions(hand);
            System.out.println(Arrays.asList(suits));
        
            assertEquals(Suit.values()[i], suits[0]);
        }

        {
            Card [] hand = new Card[] { 
                    new Card(Rank.EIGHT, Suit.DIAMONDS), 
                    new Card(Rank.NINE, Suit.DIAMONDS), 
                    new Card(Rank.TEN, Suit.DIAMONDS), 
                    new Card(Rank.EIGHT, Suit.HEARTS), 
                    new Card(Rank.EIGHT, Suit.HEARTS), 
                    new Card(Rank.EIGHT, Suit.CLUBS) 
            };
    
            Suit [] suits = Kaiser.getBestTrumpOptions(hand);
            assertEquals(Suit.DIAMONDS, suits[0]);
            assertEquals(Suit.HEARTS, suits[1]);
            assertEquals(Suit.CLUBS, suits[2]);
        }


    }
    
    public void test_computeBidOptions() {
        
        Card [] hand = new Card[] { 
                new Card(Rank.EIGHT, Suit.DIAMONDS), 
                new Card(Rank.NINE, Suit.DIAMONDS), 
                new Card(Rank.TEN, Suit.CLUBS), 
                new Card(Rank.EIGHT, Suit.HEARTS), 
                new Card(Rank.ACE, Suit.CLUBS), 
                new Card(Rank.QUEEN, Suit.DIAMONDS), 
                new Card(Rank.TEN, Suit.SPADES), 
                new Card(Rank.FIVE, Suit.HEARTS), 
        };
        
        Bid [] options;
        //Bid [] options = Kaiser.computeBidOptions(null, false, hand);
        //System.out.println(Arrays.asList(options));
        //options = Kaiser.computeBidOptions(null, true, hand);
        //System.out.println(Arrays.asList(options));
        options = Kaiser.computeBidOptions(new Bid(10, Suit.HEARTS), false, hand);
        System.out.println(Arrays.asList(options));
        options = Kaiser.computeBidOptions(new Bid(10, Suit.HEARTS), true, hand);
        System.out.println(Arrays.asList(options));
        options = Kaiser.computeBidOptions(new Bid(10, Suit.NOTRUMP), false, hand);
        System.out.println(Arrays.asList(options));
        options = Kaiser.computeBidOptions(new Bid(10, Suit.NOTRUMP), true, hand);
        System.out.println(Arrays.asList(options));
    }
}

File: Kaiser/test/cc/game/kaiser/core/KaiserTest.kt
package cc.game.kaiser.core

import cc.game.kaiser.ai.PlayerBot
import cc.lib.game.Utils
import junit.framework.TestCase
import java.io.File
import java.io.PrintWriter
import java.io.StringWriter
import java.util.*

class KaiserTest : TestCase() {
	public override fun setUp() {
		Kaiser.DEBUG_ENABLED = true
		println("Start Test: $name")
	}

	@Throws(Exception::class)
	fun testSerializeTeam() {
		val team = Team()
		team.bid = NO_BID
		team.name = "scrubs"
		team.roundPoints = 10000
		team.totalPoints = 98320982

		//PrintWriter out = new PrintWriter(System.out);
		val out = StringWriter()
		team.serialize(PrintWriter(out))
		println(out.toString())
		val t2 = Team()
		t2.deserialize(out.toString())
		assertEquals(t2.totalPoints, team.totalPoints)
		assertEquals(t2.roundPoints, team.roundPoints)
		assertEquals(t2.name, team.name)
		assertEquals(t2, team)
	}

	@Throws(Exception::class)
	fun testKaiser() {
		var saved = false
		val k = Kaiser()
		k.setPlayer(0, PlayerBot("Joe Blow"))
		k.setPlayer(1, PlayerBot("Pat Robertson"))
		k.setPlayer(2, PlayerBot("Bobbing for Apples"))
		k.setPlayer(3, PlayerBot("Sam crow"))
		while (k.state !== State.GAME_OVER) {
			if (k.state === State.NEW_ROUND && !saved && Math.random() * 10 < 2) {
				k.saveToFile(File("resources/kaisertest.txt"))
				saved = true
			}
			k.runGame()
		}
	}

	@Throws(Exception::class)
	fun testRestoreGame() {
		Kaiser.DEBUG_ENABLED = true
		Utils.setDebugEnabled()
		var k = Kaiser()
		k.setPlayer(0, PlayerBot("Joe Blow"))
		k.setPlayer(1, PlayerBot("Pat Robertson"))
		k.setPlayer(2, PlayerBot("Bobbing for Apples"))
		k.setPlayer(3, PlayerBot("Sam crow"))
		while (k.state !== State.GAME_OVER) {
			k.saveToFile(File("resources/test.txt"))
			k = Kaiser()
			k.loadFromFile(File("resources/test.txt"))
			k.runGame()
			println("State after run: " + k.state)
		}
	}

	@Throws(Exception::class)
	fun testRestoreGame2() {
		val k = Kaiser()
		k.loadFromFile(File("resources/savegame.txt"))
		val p = k.getPlayer(0)
		assertEquals("Chris", p.name)
		assertEquals(7, p.numCards)
		assertEquals(1, k.getTeam(1).roundPoints)
		assertEquals(10, k.getTeam(1).roundPoints)
		var cnt = 1000
		while (cnt-- > 0 && k.state !== State.GAME_OVER) {
			k.runGame()
		}
		assert(k.state === State.GAME_OVER)
	}

	@Throws(Exception::class)
	fun testSerialize() {
		val k = Kaiser()
		k.loadFromFile(File("resources/kaisertest.txt"))
		while (k.state !== State.GAME_OVER) {
			k.runGame()
		}
	}

	fun testGetTrickWinnerIndex() {
		val aceClubs = Card(Rank.ACE, Suit.CLUBS)
		val kngDmnds = Card(Rank.KING, Suit.DIAMONDS)
		val qunHrts = Card(Rank.QUEEN, Suit.HEARTS)
		val jckSpds = Card(Rank.JACK, Suit.SPADES)
		val jckHrts = Card(Rank.JACK, Suit.HEARTS)
		val jckClubs = Card(Rank.JACK, Suit.CLUBS)
		val jckDmnds = Card(Rank.JACK, Suit.DIAMONDS)
		val qunDmnds = Card(Rank.QUEEN, Suit.DIAMONDS)
		val tenDmnds = Card(Rank.TEN, Suit.DIAMONDS)

		// test when all the same, no trump
		doTestTrick(arrayOf(jckSpds, jckHrts, jckClubs, jckDmnds), Suit.NOTRUMP, 0, jckSpds)
		doTestTrick(arrayOf(jckSpds, jckHrts, jckClubs, jckDmnds), Suit.NOTRUMP, 1, jckHrts)
		doTestTrick(arrayOf(jckSpds, jckHrts, jckClubs, jckDmnds), Suit.NOTRUMP, 2, jckClubs)
		doTestTrick(arrayOf(jckSpds, jckHrts, jckClubs, jckDmnds), Suit.NOTRUMP, 3, jckDmnds)

		// test when all the same, has trump
		for (i in 0..3) {
			doTestTrick(arrayOf(jckSpds, jckHrts, jckClubs, jckDmnds), Suit.SPADES, i, jckSpds)
			doTestTrick(arrayOf(jckSpds, jckHrts, jckClubs, jckDmnds), Suit.HEARTS, i, jckHrts)
			doTestTrick(arrayOf(jckSpds, jckHrts, jckClubs, jckDmnds), Suit.CLUBS, i, jckClubs)
			doTestTrick(arrayOf(jckSpds, jckHrts, jckClubs, jckDmnds), Suit.DIAMONDS, i, jckDmnds)
		}

		// test when all different, no trump
		doTestTrick(arrayOf(aceClubs, kngDmnds, qunHrts, jckSpds), Suit.NOTRUMP, 0, aceClubs)
		doTestTrick(arrayOf(aceClubs, kngDmnds, qunHrts, jckSpds), Suit.NOTRUMP, 1, kngDmnds)
		doTestTrick(arrayOf(aceClubs, kngDmnds, qunHrts, jckSpds), Suit.NOTRUMP, 2, qunHrts)
		doTestTrick(arrayOf(aceClubs, kngDmnds, qunHrts, jckSpds), Suit.NOTRUMP, 3, jckSpds)
		doTestTrick(arrayOf(aceClubs, qunHrts, jckSpds, jckHrts), Suit.NOTRUMP, 0, aceClubs)
		doTestTrick(arrayOf(aceClubs, qunHrts, jckSpds, jckHrts), Suit.NOTRUMP, 1, qunHrts)
		doTestTrick(arrayOf(aceClubs, qunHrts, jckSpds, jckHrts), Suit.NOTRUMP, 2, jckSpds)
		doTestTrick(arrayOf(aceClubs, qunHrts, jckSpds, jckHrts), Suit.NOTRUMP, 3, qunHrts)

		// test when all different, has trump
		val suits = arrayOf(Suit.CLUBS, Suit.DIAMONDS, Suit.HEARTS, Suit.SPADES)
		val cards = arrayOf(aceClubs, kngDmnds, qunHrts, jckSpds)
		var index = 0
		for (trump in suits) {
			doTestTrick(cards, trump, 0, cards[index])
			doTestTrick(cards, trump, 1, cards[index])
			doTestTrick(cards, trump, 2, cards[index])
			doTestTrick(cards, trump, 3, cards[index])
			index++
		}

		// test all the same suit, no trump
		for (i in 0..3) doTestTrick(arrayOf(kngDmnds, qunDmnds, jckDmnds, tenDmnds), Suit.NOTRUMP, i, kngDmnds)
		for (i in 0..3) doTestTrick(arrayOf(kngDmnds, qunDmnds, jckDmnds, tenDmnds), Suit.DIAMONDS, i, kngDmnds)
		for (i in 0..3) doTestTrick(arrayOf(kngDmnds, qunDmnds, jckDmnds, tenDmnds), Suit.SPADES, i, kngDmnds)
	}

	fun doTestTrick(trick: Array<Card>, trump: Suit?, startIndex: Int, expected: Card?) {
//		val index: Int = Kaiser.trgetrickWinnerIndex
//		assertEquals(expected, trick[index])
	}

	/*
    public void testParseNextLineElement() {
        String line = "PLAYER:0 cc.game.kaiser.swing.SwingPlayerUser \"Chris\"";

        assertEquals("PLAYER", Kaiser.parseNextLineElement(line, 0));
        assertEquals("0", Kaiser.parseNextLineElement(line, 1));
        assertEquals("cc.game.kaiser.swing.SwingPlayerUser", Kaiser.parseNextLineElement(line, 2));
        assertEquals("Chris", Kaiser.parseNextLineElement(line, 3));

    }
   */
	fun test_getBestTrumpOptions() {
		val k = Kaiser()
		for (i in 0..3) {
			val hand = listOf(k.getCard(Rank.ACE, Suit.values()[i]))
			val suits = Kaiser.getBestTrumpOptions(hand)
			println(Arrays.asList(*suits))
			assertEquals(Suit.values()[i], suits[0])
		}
		run {
			val hand = listOf(
				Card(Rank.EIGHT, Suit.DIAMONDS),
				Card(Rank.NINE, Suit.DIAMONDS),
				Card(Rank.TEN, Suit.DIAMONDS),
				Card(Rank.EIGHT, Suit.HEARTS),
				Card(Rank.EIGHT, Suit.HEARTS),
				Card(Rank.EIGHT, Suit.CLUBS)
			)
			val suits = Kaiser.getBestTrumpOptions(hand)
			assertEquals(Suit.DIAMONDS, suits[0])
			assertEquals(Suit.HEARTS, suits[1])
			assertEquals(Suit.CLUBS, suits[2])
		}
	}

	fun test_computeBidOptions() {
		val hand = listOf(
			Card(Rank.EIGHT, Suit.DIAMONDS),
			Card(Rank.NINE, Suit.DIAMONDS),
			Card(Rank.TEN, Suit.CLUBS),
			Card(Rank.EIGHT, Suit.HEARTS),
			Card(Rank.ACE, Suit.CLUBS),
			Card(Rank.QUEEN, Suit.DIAMONDS),
			Card(Rank.TEN, Suit.SPADES),
			Card(Rank.FIVE, Suit.HEARTS))
		var options: Array<Bid>
		//Bid [] options = Kaiser.computeBidOptions(null, false, hand);
		//System.out.println(Arrays.asList(options));
		//options = Kaiser.computeBidOptions(null, true, hand);
		//System.out.println(Arrays.asList(options));
		options = Kaiser.computeBidOptions(Bid(10, Suit.HEARTS), false, hand)
		println(Arrays.asList(*options))
		options = Kaiser.computeBidOptions(Bid(10, Suit.HEARTS), true, hand)
		println(Arrays.asList(*options))
		options = Kaiser.computeBidOptions(Bid(10, Suit.NOTRUMP), false, hand)
		println(Arrays.asList(*options))
		options = Kaiser.computeBidOptions(Bid(10, Suit.NOTRUMP), true, hand)
		println(Arrays.asList(*options))
	}
}